### PR TITLE
Fixed #25912 -- Added binary left/right shift operators to F expressions.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/25912
Updated from https://github.com/django/django/pull/5815.

### CODE REVIEW

File: tests/expressions/tests.py [hunk 1]
```diff
@@ -741,6 +741,14 @@ def test_lefthand_bitwise_and(self):
         self.assertEqual(Number.objects.get(pk=self.n.pk).integer, 40)
         self.assertEqual(Number.objects.get(pk=self.n.pk).float, Approximate(15.500, places=3))
 
+    def test_lefthand_bitwise_left_shift_operator(self):
+        Number.objects.filter(pk=self.n.pk).update(integer=F('integer').bitleftshift(2))
+        self.assertEqual(Number.objects.get(pk=self.n.pk).integer, 168)
+
+    def test_lefthand_bitwise_right_shift_operator(self):
+        Number.objects.filter(pk=self.n.pk).update(integer=F('integer').bitrightshift(2))
```
Comment (felixxm) [inline tests/expressions/tests.py]: I think that we need additional tests for negative values, at least for binary right shift operator:

```python
>>> -123>>2
-31
>>> 123>>2
30
```

Comment (timgraham) [inline tests/expressions/tests.py]: Good idea, I added that and it didn't require changes to the implementation. I'm not certain if `FLOOR(...)` is correct but tests are passing.

**Comment by timgraham** [comment]
@adamchainz, for the negative values cases, MySQL gives:
```
(1264, "Out of range value for column 'the_integer' at row 2")
AssertionError: 9223372036854775807 != -168
AssertionError: 4611686018427387893 != -11
```
Any idea about that behavior?

**Comment by felixxm** [comment]
MySQL binary left and right shift operators return result as a unsigned 64-bit integer (see [doc](http://dev.mysql.com/doc/refman/5.7/en/bit-functions.html#operator_left-shift)). I think that we should use the same approach as in Oracle backend, i.e.:
```diff
diff --git a/django/db/backends/mysql/operations.py b/django/db/backends/mysql/operations.py
index dd9f56e..2f8c1d9 100644
--- a/django/db/backends/mysql/operations.py
+++ b/django/db/backends/mysql/operations.py
@@ -202,8 +202,13 @@ class DatabaseOperations(BaseDatabaseOperations):
         """
         MySQL requires special cases for ^ operators in query expressions
         """
+        lhs, rhs = sub_expressions
         if connector == '^':
             return 'POW(%s)' % ','.join(sub_expressions)
+        elif connector == '<<':
+            return '(%(lhs)s * POW(2, %(rhs)s))' % {'lhs': lhs, 'rhs': rhs}
+        elif connector == '>>':
+            return 'FLOOR(%(lhs)s / POW(2, %(rhs)s))' % {'lhs': lhs, 'rhs': rhs}
         return super(DatabaseOperations, self).combine_expression(connector, sub_expressions)
```

**Comment by timgraham** [comment]
Thanks for the tip, Mariusz.

File: django/db/backends/oracle/operations.py [hunk 2]
```diff
@@ -437,13 +437,17 @@ def adapt_timefield_value(self, value):
 
     def combine_expression(self, connector, sub_expressions):
         "Oracle requires special cases for %% and & operators in query expressions"
```
Comment (charettes) [inline django/db/backends/oracle/operations.py]: Might be worth removing or adjusting the docstring?

**Comment by timgraham** [comment]
bulidbot, test on oracle.

File: django/db/backends/oracle/operations.py [hunk 3]
```diff
@@ -436,14 +436,17 @@ def adapt_timefield_value(self, value):
...
             return 'MOD(%s)' % ','.join(sub_expressions)
         elif connector == '&':
             return 'BITAND(%s)' % ','.join(sub_expressions)
         elif connector == '|':
-            lhs, rhs = sub_expressions
             return 'BITAND(-%(lhs)s-1,%(rhs)s)+%(lhs)s' % {'lhs': lhs, 'rhs': rhs}
+        elif connector == '<<':
+            return '(%(lhs)s * POWER(2, %(rhs)s))' % {'lhs': lhs, 'rhs': rhs}
+        elif connector == '>>':
+            return 'FLOOR(%(lhs)s / POWER(2, %(rhs)s))' % {'lhs': lhs, 'rhs': rhs}
```
Comment (felixxm) [inline django/db/backends/oracle/operations.py]: I think that I found complex solution for all binary operators in `MySQL` that return an unsigned 64-bit integer. We can simple convert return value to `SIGNED` integer. I don't know why, but it doesn't work for right shift operator.
```diff
--- a/django/db/backends/mysql/operations.py
+++ b/django/db/backends/mysql/operations.py
@@ -202,8 +202,14 @@ class DatabaseOperations(BaseDatabaseOperations):
+        lhs, rhs = sub_expressions
         if connector == '^':
             return 'POW(%s)' % ','.join(sub_expressions)
+        # MySQL's binary operators return an unsigned 64-bit integer.
+        elif connector in ['<<', '|', '&']:
+            return 'CONVERT(%s, SIGNED)' % connector.join(sub_expressions)
+        elif connector == '>>':
+            return 'FLOOR(%(lhs)s / POW(2, %(rhs)s))' % {'lhs': lhs, 'rhs': rhs}
         return super(DatabaseOperations, self).combine_expression(connector, sub_expressions)
```

Comment (timgraham) [inline django/db/backends/oracle/operations.py]: Would it be better as a separate PR since |/& are not otherwise modified in this commit (also some tests are needed to show what changed for them?).

Comment (felixxm) [inline django/db/backends/oracle/operations.py]: You're right. I prepared separate PR #7787.

Comment (timgraham) [inline django/db/backends/oracle/operations.py]: I rebased this PR and merging yours.
